### PR TITLE
Add downlevel dts master

### LINKED ISSUE #9463: TS1086: An accessor cannot be declared in an ambient context
**Describe the bug**
Upgraded to 5.3.3 just now and our Angular build is failing with the following:

```
[ng]
[ng] ERROR in ../../node_modules/@storybook/channels/dist/index.d.ts:25:9 - error TS1086: An accessor cannot be declared in an ambient context.
[ng]
[ng] 25     get hasTransport(): boolean;
[ng]            ~~~~~~~~~~~~
```

From a quick google, it appears other repos are facing the same issue due to the use of TypeScript 3.7.0 - but Angular is restricted to <3.5.0

**Similar errors**
https://github.com/nestjs/nest/issues/3513
https://github.com/googleapis/node-gtoken/issues/244

I believe the breaking change come from TypeScript directly https://github.com/microsoft/TypeScript/issues/33939 - but until a fix is released, the `index.d.ts` file will need regenerating with `npm i typescript@~3.6.0` (swap `^` for `~`)

**System:**
```
    OS: macOS Mojave 10.14.6
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
  Binaries:
    Node: 12.14.1 - /usr/local/bin/node
    npm: 6.13.4 - /usr/local/bin/npm
  Browsers:
    Chrome: 79.0.3945.117
    Safari: 13.0.4
  npmPackages:
    @storybook/addon-actions: ^5.3.3 => 5.3.3
    @storybook/addon-knobs: ^5.3.3 => 5.3.3
    @storybook/addon-links: ^5.3.3 => 5.3.3
    @storybook/addon-notes: ^5.3.3 => 5.3.3
    @storybook/addon-storyshots: ^5.3.3 => 5.3.3
    @storybook/addons: ^5.3.3 => 5.3.3
    @storybook/angular: ^5.3.3 => 5.3.3
    @storybook/cli: ^5.3.3 => 5.3.3
```

### PR DESCRIPTION
**Pull request by lychyi**
Issue: Fixes #9463, an alternative to #9826 

While there are incompatible declaration outputs between TypeScript 3.7+ and TS3.5, there is a workaround to use the latest TypeScript version while still maintaining compatible `d.ts` files. This is accomplished by using [downlevel-dts](https://github.com/sandersn/downlevel-dts) to generate alternative typings for users on a certain version of TypeScript.

In this PR, any users who are on TS3.5 and below, TypeScript will automatically look at the `package.json` for a field called `typesVersions` and utilize the typings in that folder via a mapping.
```
  "typesVersions": {
    "<=3.5": { "*": ["ts3.5/*"] }
  },
```
## What I did

I modified the `compile-tsc.js` script to use `downlevel-dts` and generate typings from the `dist` folder to a `ts3.5/dist` folder. Then I added the above snippet to each `package.json` with a corresponding `types` field. I'm assuming other `package.json`s without `types` are not actually exporting any TypeScript files.

Refer to the comments in #9463 for more detail.

## How to test

I tested it by manually altering `node_modules` with this output in `@storybook/channels` inside my own project that suffers from #9463. The project indeeds builds correctly now with TypeScript 3.5.3 and Storybook 5.3.12.

I also verified that the folder `lib/channels/ts3.5/dist/index.d.ts` has this line:

```
readonly hasTransport: boolean;
```
instead of 
```
get hasTransport(): boolean;
```

I'd probably like a second set of eyes on it to make sure it works and also help out on any automated tests(if we think it's worth the effort).

### CODE REVIEW

File: scripts/compile-tsc.js
Comment (lychyi) [inline scripts/compile-tsc.js]: After trying to add parameters to make this configurable, I figured it was cleaner and more rational to just hardcode it.

The src and target directories for `downlevel-dts` must work in tandem with the `package.json` so passing it through as a parameter doesn't seem to add any value.

Let me know if there's some other flexibility needs I need to account for!

Comment (lychyi) [reply scripts/compile-tsc.js]: Also, for those wondering "Why ts3.5 as the target dir name?" (okay maybe no one is wondering that...)

`downlevel-dts` doesn't target a TS version. It just makes the latest TS (3.8 in this case) compatible with older versions via transformations of a d.ts file. I had to do some digging into what @sandersn was targeting since the [repo](https://github.com/sandersn/downlevel-dts) is strangely quiet about it. The README recommends `ts3.4` but that didn't seem right to me since TS3.5 users would benefit as well.

So I read [Twitter](https://twitter.com/sanders_n/status/1217865246557126656) instead and it looks like it ships types for TS3.5 or earlier. Thus, the naming for the target directory: `ts3.5`.

Comment (shilman) [reply scripts/compile-tsc.js]: 😂

Comment (gaetanmaisse) [reply scripts/compile-tsc.js]: I think we should output in `dist/ts3.5` instead of `ts3.5/dist`. Some reasons for this thought:
 - Everything that is "compiled/transpiled" is in a single folder: `dist`
 - It's a common practice to have the result of a compilation stage output into `dist`
 - `dist/**` is listed in `files` attributes of `package.json` (https://github.com/storybookjs/storybook/pull/9847/commits/038b7f9d76ae8de5c324d56dbf7b5ad876ea4dd5#diff-353aed04b1a7ddb0879e79df26422780R29) meaning it is packaged inside NPM package. 
⚠️ If we keep `ts3.5/dist` we must add it to the `files` attributes of all packages.  

@lychyi what is your opinion on that?

**Review by gaetanmaisse — COMMENTED** [review]
You have updated `package.json` of all `libs` and `addons`, I think we need to do the same for all `/apps` packages that are written in TS.

File: addons/a11y/package.json
Comment (gaetanmaisse) [inline addons/a11y/package.json]: Maybe to update, after discussing https://github.com/storybookjs/storybook/pull/9847#discussion_r379344314

**Comment by gaetanmaisse** [comment]
@lychyi  Thanks for this PR 😍 great work! 👍
We just need to discuss the output folder https://github.com/storybookjs/storybook/pull/9847#discussion_r379344314 ;)

**Comment by shilman** [comment]
Y'all are so on top of this stuff. Great work!!! 🙇🙇🙇

File: scripts/compile-tsc.js
Comment (sandersn) [reply scripts/compile-tsc.js]: downlevel-dts started off as a tiny one-off piece of code, so I've been slower than I should to add real documentation. I just added some, so the README has substantially more detail now.

**Comment by lychyi** [comment]
@gaetanmaisse Good points, I just followed the example on downlevel-dts so I had to think about it for a sec. This was the most compelling reason I could think of: 

`ts3.5/dist` is essentially mirroring whatever was in `dist` in another location. If we do `dist/ts3.5`, in the _very_ low chance that someone names a src folder `ts3.5` you'll essentially clobber what's in there when you compile and downlevel TS into `dist`. So the downside in this case is that `ts3.5` becomes a reserved folder name that you can't use, it's not a deal breaker but it's just a mental nugget that you have to keep in mind. 

And yes, good call on the `files`. 😅 and setting `typesVersions` for `app/`. I did a find and replace and I'm baffled how I missed that.

**Comment by gaetanmaisse** [comment]
@ndelangen @kroeder what do you think about downlevel-dts' output folder ⬆️ https://github.com/storybookjs/storybook/pull/9847#discussion_r379344314 and https://github.com/storybookjs/storybook/pull/9847#issuecomment-586507675?

**Comment by lychyi** [comment]
✏️ Just writing down a mental note here ✏️ 

This PR still may need the `files` attribute altered in `package.json` before merging. I'll do it once the decision is made about where we downlevel things to.

**Comment by gaetanmaisse** [comment]
I discussed with @ndelangen and he hasn't a strong preference on it, he would use `ts3.5/dist` as that's the proposed way from the tools. So I think we can agree on this! 

@lychyi can you tag me when you'll add a commit about the `files` attribute? 

🗒 Reminder: Add these changes to `next` branch too.

**Comment by lychyi** [comment]
@gaetanmaisse `files` attribute has been updated.

Note: I added this to every `package.json` that specifies that it ships `*.d.ts` in `files` even though the package may not have `d.ts` files to ship. I'm still debating if this is the right call but I figured if we're specifying that it could ship `d.ts` files in the future, we would need to ship downleveled types too so I made it consistent across the project so we don't have to account for it in the future.

For example, if someone wants to create a new package, do they copy and paste from an existing `package.json` or is there a script that generates one? If it's the former, then having it in every `package.json` would make sense.

Maintainers: feel free to modify the PR if that wasn't the right call or I can do it as soon as you let me know.

**Review by gaetanmaisse — APPROVED** [review]

**Comment by gaetanmaisse** [comment]
@lychyi LGTM! I just added a commit with some `typesVersions` you missed. Or was it intentional? 

Tested locally:
  - Reproduce https://github.com/storybookjs/storybook/issues/9463
  - Use packages of this branch with `yarn link`
  - Everything was 🟢

@shilman @ndelangen  don't know if someone else wants to take a look 👨🏻‍💻  If not, I think it can be merged, what's the process, merge on `master` and then `master` is merged in `next`? or merged on `next` and then cherry-picked on `master`?

**Comment by ndelangen** [comment]
@gaetanmaisse This PR targets master, so let's merge this to master first

I'm unsure if cherry-picking this is easier then creating an additional PR against next?

**Review by ndelangen — APPROVED** [review]

**Comment by lychyi** [comment]
@gaetanmaisse Ah, I think I wrongly assumed that only packages with the `types` attribute shipped typings to downlevel. Good call on this.

**Comment by gaetanmaisse** [comment]
@lychyi I think it was a valid assumption but there are suspicious things with these packages I will take a look at them 🔍🕵

**Comment by gaetanmaisse** [comment]
@lychyi Thanks for your work 🎉 🎉 🙇 

I will cherry-pick and create the PR for `next` ASAP.

**Comment by lychyi** [comment]
🎉Thanks for the assist! ❤️ Lovin' the project and are more than happy to help. @gaetanmaisse @ndelangen @shilman

**Comment by shilman** [comment]
@lychyi @gaetanmaisse @ndelangen I just realized that this was merged into `master` and the second PR is going into `next`! 🙀

I just released the second PR as part of https://github.com/storybookjs/storybook/releases/tag/v6.0.0-alpha.14 

Since this feels like a biggish change, can you please QA that release a bit and let me know if it looks good? Once I get your go-ahead, I'll release this change on 5.3.x.

Thanks!! 🙏🙏🙏

**Comment by gaetanmaisse** [comment]
🛑 It looks like there is a bug with [Addon Centered](https://github.com/storybookjs/storybook/tree/next/addons/centered) because it's the only package that export types from "root" dir. Here is the fix: https://github.com/storybookjs/storybook/pull/9907
Everything else looks good.

**Pull request merged**
